### PR TITLE
Allow to call arbitrary runtime apis using RelayChainInterface

### PR DESCRIPTION
**Pull request by tmpolaczyk**
When using the relay chain though a `Arc<dyn RelayChainInterface>` there is no way to call arbitrary runtime apis. Both implementations of that trait allow this, so it feels natural to expose this functionality in the trait.

This PR adds a `call_runtime_api` method to RelayChainInterface trait, and a separate function also named `call_runtime_api` which allows the caller to specify the input and output types, as opposed to having to encode them. This generic function cannot be part of the trait because a `dyn Trait` object cannot have generic methods.

### CODE REVIEW

File: cumulus/client/relay-chain-inprocess-interface/src/lib.rs [hunk 1]
```diff
@@ -176,6 +176,29 @@ impl RelayChainInterface for RelayChainInProcessInterface {
 		Ok(self.backend.blockchain().info().finalized_hash)
 	}
 
+	async fn call_remote_runtime_function_encoded(
+		&self,
+		method_name: &'static str,
+		hash: PHash,
+		payload: &[u8],
+	) -> RelayChainResult<Vec<u8>> {
+		use sp_api::{CallApiAt, CallApiAtParams, CallContext, __private::Extensions};
```
Comment (bkchr) [inline cumulus/client/relay-chain-inprocess-interface/src/lib.rs:185]: The module is called `__private` out of a reason ;)

File: cumulus/client/relay-chain-inprocess-interface/src/lib.rs [hunk 2]
```diff
@@ -176,6 +176,29 @@ impl RelayChainInterface for RelayChainInProcessInterface {
...
+		use sp_state_machine::OverlayedChanges;
+		use std::cell::RefCell;
+
+		let overlayed_changes = OverlayedChanges::default();
+
+		Ok(self.full_client.call_api_at(CallApiAtParams {
+			at: hash,
+			function: method_name,
+			arguments: payload.to_vec(),
+			overlayed_changes: &RefCell::new(overlayed_changes),
```
Comment (bkchr) [inline cumulus/client/relay-chain-inprocess-interface/src/lib.rs:195]:
```suggestion
			overlayed_changes: &RefCell::new(Default::default()),
```

File: cumulus/client/relay-chain-interface/src/lib.rs [hunk 3]
```diff
@@ -364,3 +380,18 @@ where
 		(**self).version(relay_parent).await
 	}
 }
+
+pub async fn call_remote_runtime_function<R>(
```
Comment (bkchr) [inline cumulus/client/relay-chain-interface/src/lib.rs:384]: Do we need this?

File: cumulus/client/relay-chain-interface/src/lib.rs [hunk 4]
```diff
@@ -117,6 +117,13 @@ pub trait RelayChainInterface: Send + Sync {
 	/// Get the hash of the finalized block.
 	async fn finalized_block_hash(&self) -> RelayChainResult<PHash>;
 
+	async fn call_remote_runtime_function_encoded(
```
Comment (bkchr) [inline cumulus/client/relay-chain-interface/src/lib.rs:120]: Missing docs. 
```suggestion
	async fn call_runtime_api(
```

File: cumulus/client/relay-chain-interface/src/lib.rs (hunk 3 above)
Comment (tmpolaczyk) [reply cumulus/client/relay-chain-interface/src/lib.rs]: It saves the caller from having to manually do the `Decode::decode`. I can remove it if you want, but I think it's useful.

File: cumulus/client/relay-chain-interface/src/lib.rs [hunk 5]
```diff
@@ -364,3 +381,18 @@ where
 		(**self).version(relay_parent).await
 	}
 }
+
+pub async fn call_runtime_api<R>(
```
Comment (bkchr) [inline cumulus/client/relay-chain-interface/src/lib.rs:385]: Still missing docs.

**Comment by tmpolaczyk** [comment]
@bkchr solved your comments, anything else missing?

**Comment by girazoki** [comment]
@bkchr can you help us get the CI passing? Just from a first glance I dont see anything wrong for prdoc but maybe I am missing something...,

**Review by bkchr — APPROVED** [review]

**Review by skunert — APPROVED** [review]

**Pull request merged**